### PR TITLE
mgr/cephadm: allow to map existing daemons to osdspecs

### PR DESCRIPTION
**Pull request by jschmid1**
Signed-off-by: Joshua Schmid <jschmid@suse.de>

with https://tracker.ceph.com/issues/44755 finally being handled, this enables us to implement the `limit` key in osdspecs/drivegroups properly.

Fixes: https://tracker.ceph.com/issues/44888

**Depends on**

- [x] #36388 
- [x] #36541

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/python-common/ceph/deployment/drive_selection/selector.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
 import logging
 
+from orchestrator import DaemonDescription
```
Comment (jschmid1) [inline src/python-common/ceph/deployment/drive_selection/selector.py:3]:
```
ceph/deployment/drive_selection/selector.py:3: in <module>
    from orchestrator import DaemonDescription
E   ModuleNotFoundError: No module named 'orchestrator'
```

resolves locally for me.. How to resolve that properly?

Comment (sebastian-philipp) [reply src/python-common/ceph/deployment/drive_selection/selector.py:3]: you should not import MGRs modules from python-common. note that python-common code will be imported by ceph-volume

File: src/python-common/ceph/deployment/drive_selection/selector.py [hunk 2]
```diff
@@ -68,7 +83,7 @@ def _limit_reached(device_filter, len_devices,
         """
         limit = device_filter.limit or 0
 
-        if limit > 0 and len_devices >= limit:
+        if limit > 0 and (len_devices >= limit or len(self.osds_for_spec) >= limit):
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/drive_selection/selector.py:86]: can you move this into a different commit?

Comment (jschmid1) [reply src/python-common/ceph/deployment/drive_selection/selector.py:84]: done

File: src/python-common/ceph/deployment/drive_selection/selector.py (hunk 1 above)
Comment (jschmid1) [reply src/python-common/ceph/deployment/drive_selection/selector.py]: done

**Comment by jschmid1** [comment]
jenkins retest this please

File: src/python-common/ceph/deployment/drive_selection/selector.py [hunk 3]
```diff
@@ -34,6 +37,17 @@ def __init__(self,
...
+    def find_osds_in_spec(self) -> List["DaemonDescription"]:
+        osds: List["DaemonDescription"] = []
+        if not self.daemons:
+            return osds
+        for host, dm in self.daemons:
+            for name, dd in dm.items():
+                if dd.daemon_type == 'osd':
+                    if dd.osdspec_affinity == self.spec.service_id:
+                        osds.append(dd)
+        return osds
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/drive_selection/selector.py:49]: What do you think of making this a static method of `DaemonDescription`? would that be ok for you?

Comment (jschmid1) [reply src/python-common/ceph/deployment/drive_selection/selector.py:49]: Wouldn't that mean that we need to import `DaemonDescription` in the `python-common` realm?

Comment (jschmid1) [reply src/python-common/ceph/deployment/drive_selection/selector.py]: There's that: https://github.com/ceph/ceph/pull/35945/commits/b39420361b655d20ccf87cda83dab45251e30f75

But I feel this is not what you had in mind.

File: src/python-common/ceph/deployment/drive_selection/selector.py [hunk 4]
```diff
@@ -34,6 +37,17 @@ def __init__(self,
             self._db = self.assign_devices(self.spec.db_devices)
             self._journal = self.assign_devices(self.spec.journal_devices)
 
+    def find_osds_in_spec(self):
+        # type: () -> List["DaemonDescription"]
+        osds = []  # type: List["DaemonDescription"]
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/drive_selection/selector.py:42]:
```
_________________ ceph/deployment/drive_selection/selector.py __________________
20: error: Name 'DaemonDescription' is not defined
40: error: Name 'DaemonDescription' is not defined
42: error: Name 'DaemonDescription' is not defined
```

I fear, this is'n going to work this way. What about just cut&paste this method into mgr/orchestrator/_interface.py?

File: src/pybind/mgr/cephadm/services/osd.py [hunk 5]
```diff
@@ -119,13 +119,14 @@ def _find_inv_for_host(hostname: str, inventory_dict: dict):
         for host in matching_hosts:
             inventory_for_host = _find_inv_for_host(host, self.mgr.cache.devices)
             logger.debug(f"Found inventory for host {inventory_for_host}")
-            drive_selection = DriveSelection(drive_group, inventory_for_host)
+            drive_selection = DriveSelection(drive_group, inventory_for_host,
+                                             self.mgr.cache.get_daemons_with_volatile_status())
```
Comment (jschmid1) [inline src/pybind/mgr/cephadm/services/osd.py:123]: with https://github.com/ceph/ceph/pull/36388 this can be `self.mgr.cache.get_daemons_by_service(drive_group.service_id)`

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2020-08-19_07:38:40-rados:cephadm-wip-swagner-testing-2020-08-18-1624-distro-basic-smithi/

**Comment by sebastian-philipp** [comment]
@jan--f  Is there a way to verify that ceph-volume still works with this change?

**Comment by sebastian-philipp** [comment]
ping @jan--f

**Comment by jan--f** [comment]
I don't think so no. Once we get #34740 merged I'll add proper support for testing c-v via the teuthology cephadm task. A review would be appreciated.

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2020-09-14_13:20:29-rados:cephadm-wip-swagner3-testing-2020-09-14-1344-distro-basic-smithi/
Failure: https://tracker.ceph.com/issues/47423

**Review by sebastian-philipp — APPROVED** [review]
